File: compiler/extensions/java/runtime/test/zserio/runtime/BitPositionUtilTest.java
package zserio.runtime;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import org.junit.jupiter.api.Test;

public class BitPositionUtilTest
{
    @Test
    public void alignTo()
    {
        final long bitPosition = 5;
        assertEquals(5, BitPositionUtil.alignTo(0, bitPosition));
        assertEquals(5, BitPositionUtil.alignTo(1, bitPosition));
        assertEquals(6, BitPositionUtil.alignTo(2, bitPosition));
        assertEquals(6, BitPositionUtil.alignTo(3, bitPosition));
        assertEquals(8, BitPositionUtil.alignTo(4, bitPosition));
        assertEquals(5, BitPositionUtil.alignTo(5, bitPosition));
        assertEquals(6, BitPositionUtil.alignTo(6, bitPosition));
        assertEquals(7, BitPositionUtil.alignTo(7, bitPosition));
        assertEquals(8, BitPositionUtil.alignTo(8, bitPosition));
    }

    @Test
    public void bitsToBytes()
    {
        assertEquals(1, BitPositionUtil.bitsToBytes(8L));
        assertEquals(3, BitPositionUtil.bitsToBytes(24L));

        assertThrows(ZserioError.class, () -> BitPositionUtil.bitsToBytes(4L));
        assertThrows(ZserioError.class, () -> BitPositionUtil.bitsToBytes(9L));
    }

    @Test
    public void bytesToBits()
    {
        assertEquals(0, BitPositionUtil.bytesToBits(0L));
        assertEquals(8, BitPositionUtil.bytesToBits(1L));
        assertEquals(16, BitPositionUtil.bytesToBits(2L));
    }
}

File: compiler/extensions/cpp11/runtime/test/zserio/BitPositionUtilTest.cpp
#include "zserio/BitPositionUtil.h"
#include "zserio/CppRuntimeException.h"

#include "gtest/gtest.h"

namespace zserio
{

TEST(BitPositionUtilTest, alignTo)
{
    const size_t bitPosition = 5;
    EXPECT_EQ(5, alignTo(0, bitPosition));
    EXPECT_EQ(5, alignTo(1, bitPosition));
    EXPECT_EQ(6, alignTo(2, bitPosition));
    EXPECT_EQ(6, alignTo(3, bitPosition));
    EXPECT_EQ(8, alignTo(4, bitPosition));
    EXPECT_EQ(5, alignTo(5, bitPosition));
    EXPECT_EQ(6, alignTo(6, bitPosition));
    EXPECT_EQ(7, alignTo(7, bitPosition));
    EXPECT_EQ(8, alignTo(8, bitPosition));
}

TEST(BitPositionUtilTest, bitsToBytes)
{
    EXPECT_EQ(0, bitsToBytes(0));
    EXPECT_EQ(1, bitsToBytes(8));
    ASSERT_THROW(bitsToBytes(9), CppRuntimeException);
    EXPECT_EQ(2, bitsToBytes(16));
}

TEST(BitPositionUtilTest, bytesToBits)
{
    EXPECT_EQ(0, bytesToBits(0));
    EXPECT_EQ(8, bytesToBits(1));
    EXPECT_EQ(16, bytesToBits(2));
}

} // namespace zserio
